needless_lifetimes: false positive with impl Trait

This code compiles:

``` rust
#![feature(conservative_impl_trait)]

trait Foo {}

struct Bar {}

struct Baz<'a> {
    bar: &'a Bar,
}

impl<'a> Foo for Baz<'a> {}

impl Bar {
    fn baz<'a>(&'a self) -> impl Foo + 'a {
        Baz { bar: self }
    }
}

fn main() {}
```

But clippy gives this warning:

```
$ cargo clippy
warning: explicit lifetimes given in parameter types where they could be elided, #[warn(needless_lifetimes)] on by default
  --> src/main.rs:14:5
   |
14 |     fn baz<'a>(&'a self) -> impl Foo + 'a {
   |     ^
   |
   = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_lifetimes
```

If you apply the suggestion:

``` diff
 impl<'a> Foo for Baz<'a> {}

 impl Bar {
-    fn baz<'a>(&'a self) -> impl Foo + 'a {
+    fn baz(&self) -> impl Foo {
         Baz { bar: self }
     }
 }
```

you end with code that doesn't compile:

```
$ cargo clippy
error[E0564]: only named lifetimes are allowed in `impl Trait`, but `` was found in the type `Baz<'_>`
  --> src/main.rs:14:22
   |
14 |     fn baz(&self) -> impl Foo {
   |                      ^^^^^^^^

error: aborting due to previous error
```

I don't know if these lifetimes will become elidable in the future but, in this case, clippy should probably not raise a warning for now.

--- a/clippy_lints/src/lifetimes.rs
+++ b/clippy_lints/src/lifetimes.rs
@@ -281,6 +281,13 @@ impl<'a, 'tcx> Visitor<'tcx> for RefVisitor<'a, 'tcx> {
             TyPath(ref path) => {
                 self.collect_anonymous_lifetimes(path, ty);
             },
+            TyImplTrait(ref param_bounds) => {
+                for bound in param_bounds {
+                    if let RegionTyParamBound(_) = *bound {
+                        self.record(&None);
+                    }
+                }
+            }
             _ => (),
         }
         walk_ty(self, ty);

--- a/clippy_lints/src/lifetimes.rs
+++ b/clippy_lints/src/lifetimes.rs
@@ -287,7 +287,7 @@ impl<'a, 'tcx> Visitor<'tcx> for RefVisitor<'a, 'tcx> {
                         self.record(&None);
                     }
                 }
-            }
+            },
             _ => (),
         }
         walk_ty(self, ty);

--- a/tests/run-pass/needless_lifetimes_impl_trait.rs
+++ b/tests/run-pass/needless_lifetimes_impl_trait.rs
@@ -1,0 +1,22 @@
+#![feature(plugin)]
+#![plugin(clippy)]
+#![feature(conservative_impl_trait)]
+#![deny(needless_lifetime)]
+
+trait Foo {}
+
+struct Bar {}
+
+struct Baz<'a> {
+    bar: &'a Bar,
+}
+
+impl<'a> Foo for Baz<'a> {}
+
+impl Bar {
+    fn baz<'a>(&'a self) -> impl Foo + 'a {
+        Baz { bar: self }
+    }
+}
+
+fn main() {}

